[refactor][140] Payment 모듈 통합에 의한 로직 수정

Dianuma opened the pull request:
📝 PR 제목 규칙 : [타입][이슈번호] PR 내용

✨ 작업 설명
---
- Payment 모듈 통합에 의한 로직 수정

🔗 관련 이슈
---
- 관련 이슈 번호: #140

📋 요구 사항과 구현 내용
---
- 요구 사항
  - `SettlementEvent` 기반 정산 흐름 제거 및 `Deposit` 중심 정산 구조로 단순화
  - `Deposit` 의 `Charge` , `Withdraw` , `Payment` , `Refund` 가 실패 했을 경우 `DepositHistory` 에 실패에 대한 내역 저장
  - `DepositService` 와 `DepositHistoryService` 분리
  - 배치 처리에서 스킵 전략과 `Non-skip` 에러 처리 로직 명확화
  - 사용하지 않는 `Kafka` 이벤트 흐름 및 관련 코드 정리

- 구현 내용
  - `settlement-event Topic` 삭제 및 `SettlementEvent` 관련 로직 제거
  - `deposit-charge-event` 토픽 생성 스크립트 삭제
  - `Settlement` 엗티티 및 `OrderEvent` vo 에 `productName` 필드 추가
  - `DepositService.chargeDeposit()` 에서 `Settlement` 처리용 메서드 분리
  - yml 파일에서 `chunk-size`, `skip-limit` 값 관리하도록 설정 추가
  - Batch 스키마 생성 `Initializer` 및 `Settlement` 더미데이터 생성용 `createSettlementWithHardcodedUserCode` 추가
  - `SettlementService` 내 `SettlementEvent` 기반 처리 코드 삭제
  - `process` 체이닝을 통한 `validate` 단계와 처리 단계 분리
  - `noRollback` 적용으로 `Skip` 정책에 맞는 트랜잭션 흐름 구성
  - `SettlementErrorCapture` 로 현재 처리 `item` 을 로컬 캐시에 저장
  - `SettlementBatchStepLogger` 에 로컬 캐시에 저장된 `item` 기반으로 `Non-skip` 예외 발생 시 상세 처리 로직 추가
  - `skipListener` 에서 `skip` 된 `item` 처리 로직 구현
  - `SettlementWriter` 를 순수 DB 저장 전용으로 역할 축소
  - `CustomDepositHistoryRepositoryImpl` 정상 동작하도록 수정
  - `depositHistory` 에 대한 로직 `depositHistoryService` 로 분리
  - `DepositTransaction` 로직 템플릿 콜백 메서드 형태로 변경
  - 내부 로직 문제로 `DepositTransaction` 이 실패 했을 경우 `DepositHistory` 에 저장하기 위해서 `saveFailedHistory` 추가
  - `InvalidDepositHistoryStatusTransitionException` 삭제
 
💬 TODO 리스트
---
- [ ] 내용 1
- [ ] 내용 2

🧠 고려사항
---
- `Settlement` 트러블슈팅
  - 기존의 `reader` 와 `writer` 만 두고 모든 `DB` 작업을 `Writer` 에서 처리했을 때 발생한 문제
    - `Writer` 에서 예외가 발생하면 `Chunk` 전체가 롤백되며 `Reader` 가 처음부터 재실행되어 첫 번째 데이터가 반복 처리됨.
    - `Dirty Checking` 으로 인해 의도치 않은 업데이트가 발생함.
    - `Writer` 는 비즈니스 로직 계층이 아니며, 예외 시 `Chunk` 전체에 롤백을 전파하는 구조적 한계가 있었음.
  - `Cursor` 기반 `Reader` ( `JpaPagingItemReader` ) 로 교체했으나 동일한 문제 발생
    - `Chunk` 롤백 시 `Reader` 가 다시 실행되는 구조는 동일했으며, `pagination` 을 사용하는 기존 설계를 유지하기로 결정함.
  - `Processor` 로 비즈니스 로직을 이동했을 때 `UnexpectedRollbackException` 발생
    - `Processor` 내부 서비스 호출에서 예외가 발생하면 `Chunk` 트랜잭션이 `rollback-only` 상태가 됨.
    - `Writer` 가 정상이어도 최종 커밋 시 `rollback-only` 때문에 `UnexpectedRollbackException` 이 발생함.
  - 관련 트랜잭션을 `REQUIRES_NEW` 로 변경했을 때 정합성 붕괴 발생
    - 독립 트랜잭션으로 커밋되기 때문에 재실행 문제는 해결된 것처럼 보였으나, `Settlement` 는 `Chunk` 롤백으로 원복되는 반면 `DepositHistory` 는 롤백되지 않는 불일치가 발생함.
    - `Batch` 트랜잭션 정책을 통제할 수 없다는 치명적인 구조적 문제가 확인됨.
  - `try-catch` 제거, `skipListener` , `SettlementErrorCapture` 적용
    - 예외 발생 후 기록 및 후처리는 가능했으나 Chunk 롤백 자체를 막지 못해 근본 문제 해결은 불가능했음.
    - 유지보수성 측면에서 구조가 개선되었으므로 적용
  - `Processor` 를 `validate` 단계와 처리 단계를 분리( `Processor Chaining` )했을 때 문제 발생
    - 체인 중 하나에서 예외 발생 시 전체 `Chunk` 롤백이 일어나며 재실행 과정에서 두 `Processor` 간 상태 불일치가 발생함.
    - 검증 실패 후 재실행된 데이터가 처리 `Processor` 로 넘어가 상태 전환 오류가 반복되는 문제가 발생함.
  - `noRollback` 적용 후 문제 해결
    - `Processor` 에서 `BaseException` 이 발생해도 `Chunk` 롤백이 일어나지 않도록 설정하여 `Reader` 재실행 문제를 제거함.
    - 예외가 발생한 `item` 만 `skip` 처리되고 나머지는 정상 처리됨.
    - `UnexpectedRollbackException` 제거 및 `Settlement` 와 `DepositHistory` 간 정합성 보존.
    - `Processor chaining` 에서도 상태 불일치 문제 해소.
    - DepositHistoryType 에 CHARGE_FAILED, PAYMENT_FAILED, WITHDRAW_FAILED, SETTLEMENT_FAILED, REFUND_FAILED 추가

- `DepositHistroy` 간단한 트러블슈팅
  - `executeDepositTransaction` 내부에서 실패 처리 메서드를 직접 호출해 AOP 프록시를 거치지 못함
  - 내부 호출로 인해 `@Transactional(REQUIRES_NEW)` 트랜잭션이 시작되지 않아 실패 기록이 롤백됨
  - 로깅 AOP 또한 프록시를 우회하여 적용되지 않음
  - 트랜잭션, 로깅 모두 정상 동작시키기 위해 `DepositHistoryService` 로 히스토리 생성/저장 책임을 분리함
  - 외부 서비스 호출로 변경되면서 프록시를 타기 때문에 `REQUIRES_NEW` 트랜잭션이 정상 적용됨
  - 실패 히스토리가 항상 독립적으로 커밋되고, 로깅 AOP 도 안정적으로 동작함

<!-- AI-GENERATOR:START -->
⚙️ AI 생성 요약
---
1. OrderEvent / Settlement 모델 확장 및 SettlementEvent 제거 
 - OrderEvent에 productName 필드 추가, 팩토리 메서드들이 productName을 포함하도록 수정 
 - SettlementEvent, OrderEventType enum 삭제 및 관련 의존성 정리 
 - DepositHistory에 SETTLEMENT, *_FAILED 타입 추가 및 상태 전이 로직 단순화 

2. Deposit/Refund/Inventory 이벤트 처리 로직 개선 
 - OrderEventService.publishOrderCompletedEvents에서 productName을 포함해 OrderEvent 발행 
 - DepositService, DepositHistoryService, PaymentRefundService에서 중복 코드 제거 및 공통 처리 메서드 활용 
 - InventoryEventProducer의 decrease 메서드 제거, rollback 전용 이벤트만 발행하도록 정리 

3. DepositHistory 조회/저장 인프라 확장 
 - DepositHistoryRepository에 LockModeType.PESSIMISTIC_WRITE 추가 및 QDepositHistory 기반 커스텀 조회 구현 
 - DepositHistoryService 도입: history 조회/저장, 실패 이력 기록, duplicate/refund 타깃 검증 기능 제공 
 - DepositHistoryType에 *_FAILED 타입 추가, TransactionStatus를 COMPLETED/FAILED로 축소 

4. Deposit 설정(Setting) 배치 파이프라인 리팩터링 
 - SettingEvent 기반 Kafka 플로우 제거, Setting 엔티티 중심의 배치 프로세스로 변경 
 - SettingStepConfiguration에서 SettingProcess/Success/Validate 프로세서로 파이프라인 재구성 
 - SettingBatchStepLogger가 SettingErrorCapture, SettingRepository를 사용해 실패/성공 로깅 

5. Setting 도메인/서비스 구조 개편 
 - Setting 엔티티에 productName, totalAmount, 설정 커미션/밸런스 계산 메서드 추가 
 - SettingServiceImpl에서 Kafka 의존 제거, SettingHistoryService·DepositHistoryService와 연동해 비즈니스 처리 
 - SettingService, SettingHistoryService에 중복 참조코드 검증·히스토리 저장 등 도메인 중심 인터페이스 정의 

6. 공통/인프라 보완 (Logging, Batch Schema, Data Init 등) 
 - LogParams, LoggingAspect의 Pointcut 확장 및 logParams 메서드 주석 처리 
 - BatchSchemaInitializer, DataInitializer 보강: 초기 Deposit/Setting/Job 데이터 생성 및 DB 타입별 스크립트 실행 
 - application.yml에 batch chunk/skip 설정, custom 배치 섹션 추가 및 env 스크립트/토픽 매핑에서 settlement-event 제거
<!-- AI-GENERATOR:END -->

Dianuma committed:
- Kafka 연결이 안될 시, 재연결을 최소 5초, 최대 10초에 한번 수행하도록 변경
- settings.gradle 에서 통합한 모듈 삭제

Dianuma committed:
- Logging AOP 에 allMethodsPointcut 추가

Dianuma committed:
Merge remote-tracking branch 'origin/feat/global-147' into refactor/payment-140

Dianuma committed:
Merge remote-tracking branch 'origin/dev' into refactor/payment-140

Dianuma committed:
Merge remote-tracking branch 'origin/dev' into refactor/payment-140

Dianuma committed:
- settlement-event Topic 제거
- git action 에서 이슈 가져올 때 branch 가 아닌 PR 제목에서 가져오도록 변경
- yml 파일에 청크 사이즈 및 스킵 리미트 설정 추가
- Batch 스키마 생성 Initializer 작성
- DataInitializer 에 createSettlementWithHardcodedUserCode 추가
- DepositEventConsumer 에서 Settlement 관련 이벤트 삭제
- DepositHistory 의 TransactionStatus 변경 메서드 삭제
- DepositService.chargeDeposit 에서 Settlement 를 위한 메서드 분리
- DepositTransactionRequest 에 productName 와 orderItemCode 를 받아 referenceCode 를 만들는 메서드 추가
- @LogParams 가 Class 와 Method 둘 다 적용 되도록 변경
- Settlement 엔티티 및 OrderEvent 필드에 productName 추가
- SettlementService 에서 SettlementEvent 관련 로직 삭제
- process 체이닝을 통해서 validate 담당과 실제 처리 부를 분할
- noRollback 적용
- SettlementErrorCapture 를 통해 현재 처리중인 item 를 로컬 캐시에 저장
- 이를 기반으로 SettlementBatchStepLogger 에 Non-Skip 에러 처리 로직 추가
- skipListener 에 skip 된 item 에 대한 skip 로직 작성
- SettlementWriter 는 순수 DB 저장만 담당

Dianuma committed:
Merge remote-tracking branch 'origin/dev' into refactor/payment-140

Dianuma committed:
- CustomDepositHistoryRepositoryImpl 정상화

Dianuma committed:
- deposit-charge-event 토픽 생성 스크립트 삭제
- OrderEventType.java 삭제
- sendInventoryDecrease 삭제

98minseok commented on payment/src/main/java/com/ll/payment/deposit/model/entity/DepositHistory.java:
```diff
@@ -56,26 +56,12 @@ public static DepositHistory create(Long depositId, Long amount, Long balanceBef
-        if ( transactionStatus != TransactionStatus.PENDING ) {
+        if ( transactionStatus != TransactionStatus.COMPLETED ) {
             throw new InvalidDepositHistoryStatusTransitionException("거래 상태가 PENDING 이어야만 CANCELLED 로 변경할 수 있습니다.");
         }
```
간단한건데 에러메시지가 잘못된것 같습니다 !

Dianuma replied:
TransactionStatus 가 딱히 필요 없어보여 삭제했습니다

Dianuma replied:
의견 감사합니다.
`Deposit` 도메인 처리 중 발생하는 실패를 기록하기 위해 `DepositHistory` 관련 로직을 별도 서비스로 분리하고 템플릿 메서드 형태로 구조를 개선했습니다. 
그 과정에서 `setTransactionCancelled` 와 `InvalidDepositHistoryStatusTransitionException` 이 사라졌으므로 확인해주세요!

Dianuma committed:
- git action 변경

tigerpoint123 commented on payment/src/main/java/com/ll/payment/deposit/model/entity/DepositHistory.java:
```diff
@@ -56,26 +56,12 @@ public static DepositHistory create(Long depositId, Long amount, Long balanceBef
-        if ( transactionStatus != TransactionStatus.PENDING ) {
+        if ( transactionStatus != TransactionStatus.COMPLETED ) {
             throw new InvalidDepositHistoryStatusTransitionException("거래 상태가 PENDING 이어야만 CANCELLED 로 변경할 수 있습니다.");
         }
```
여기는 error code로 분류해도 되겠네요

Dianuma replied:
의견 감사합니다.
`Deposit` 도메인 처리 중 발생하는 실패를 기록하기 위해 `DepositHistory` 관련 로직을 별도 서비스로 분리하고 템플릿 메서드 형태로 구조를 개선했습니다. 
그 과정에서 `setTransactionCancelled` 와 `InvalidDepositHistoryStatusTransitionException` 이 사라졌으므로 확인해주세요!

98minseok reviewed (approved):
고생하셨습니다.

Dianuma committed:
- TransactionStatus 삭제

Dianuma committed:
Merge remote-tracking branch 'origin/dev' into refactor/payment-140

# Conflicts:
#	payment/src/main/resources/application.yml

Dianuma committed:
- TransactionStatus 복구 -> 상태값은 COMPLETED, FAILED 만 사용
- DepositHistoryType 에 CHARGE_FAILED, PAYMENT_FAILED, WITHDRAW_FAILED, SETTLEMENT_FAILED, REFUND_FAILED 추가
- depositHistory 에 대한 로직 depositHistoryService 로 분리
- DepositTransaction 로직 템플릿 콜백 메서드 형태로 변경
- 내부 로직 문제로 DepositTransaction 이 실패 했을 경우 DepositHistory 에 저장하기 위해서 saveFailedHistory 추가
- InvalidDepositHistoryStatusTransitionException 삭제

Dianuma committed:
- gitAction 수정

Dianuma committed:
- Deposit 에 settlement 메서드 추가

Dianuma committed:
- findByUserCode 에 대해 Lock 사용 -> 비관적 락

Dianuma committed:
- Lock 사용에 따라 트랜잭션 추가

Dianuma committed:
Merge remote-tracking branch 'origin/dev' into refactor/payment-140

Dianuma committed:
- Batch 스키마가 이미 존재할 경우 Init 생략
